board/technologic/ts7100/micro.c

```diff
@@ -0,0 +1,168 @@
+#include <stdint.h>
+#include <common.h>
+#include <asm/gpio.h>
+#include <linux/sizes.h>
+#include <asm/io.h>
+#include <common.h>
+#include <asm/arch/mx6-pins.h>
+#include <asm/mach-imx/iomux-v3.h>
+#include <i2c.h>
+#include <console.h>
+
+#include "micro.h"
+
+#define MIN_CHARGE_MV 3680
+#define MAX_CHARGE_MV 4800
+#define POWER_FAIL		IMX_GPIO_NR(5, 0)
+#define MICRO_ADDR 0x54
+
+static int micro_chip_addr;
+
+static bool read_power_fail_status(void);
+
+int micro_read(uint16_t addr, void *data, size_t size)
+{
+	struct udevice *dev;
+	int ret;
+
+	ret = i2c_get_chip_for_busnum(0, MICRO_ADDR, 2, &dev);
+	if (ret) {
+		pr_err("couldn't get i2c bus - %d\n", ret);
+		return ret;
+	}
+	return dm_i2c_read(dev, addr, data, size);
+}
+
+int micro_write(uint16_t addr, const void *data, size_t size)
+{
+	struct udevice *dev;
+	int ret = i2c_get_chip_for_busnum(0, MICRO_ADDR, 2, &dev);
+	if (ret) {
+		pr_err("couldn't get i2c bus - %d\n", ret);
+		return ret;
+	}
+
+	return dm_i2c_write(dev, addr, data, size);
+}
+
+uint8_t micro_scaps_remaining_pct(void)
+{
+	uint16_t current_voltage;
+	uint32_t voltage_range, normalized_voltage;
+	uint8_t remaining_percentage;
+
+	// Read the current supercap voltage
+	if (micro_read16_swap(MICRO_ADC_8, &current_voltage) < 0) {
+		printf("Failed to read current supercap voltage");
+		return 0;
+	}
+
+	// Calculate remaining percentage
+	if (current_voltage <= MIN_CHARGE_MV) {
+		remaining_percentage = 0;
+	} else {
+		normalized_voltage = current_voltage - MIN_CHARGE_MV;
+		voltage_range = MAX_CHARGE_MV - MIN_CHARGE_MV;
+
+		if (normalized_voltage >= voltage_range) {
+			remaining_percentage = 100;
+		} else {
+			remaining_percentage = (normalized_voltage * 100 / voltage_range);
+		}
+	}
+
+	// Ensure the remaining percentage does not exceed 100%
+	if (remaining_percentage > 100) {
+		remaining_percentage = 100;
+	}
+
+	return remaining_percentage;
+}
+
+uint16_t swap_endian16(uint16_t value)
+{
+	return (value << 8) | (value >> 8);
+}
+
+uint32_t swap_endian32(uint32_t value)
+{
+	return ((value >> 24) & 0x000000FF) | ((value >> 8) & 0x0000FF00) | ((value << 8) & 0x00FF0000) |
+	       ((value << 24) & 0xFF000000);
+}
+
+int micro_read16_swap(int addr, uint16_t *data)
+{
+	int result = micro_read(addr, (uint16_t *)data, sizeof(uint16_t));
+	if (result >= 0)
+		*data = swap_endian16(*data);
+	return result;
+}
+
+void micro_scaps_en(int en)
+{
+	uint8_t value;
+
+	micro_read8(MICRO_STATUS_FLAGS, &value);
+	value &= ~6;
```

Comment:
What is this magic number?

Reply:
This was a carryover from the original utility. From looking at the microcontroller code I dont think this is actually required. This masks the supercap enable and supercaps met minimum flags.  Removed this mask in the latest.